bpo-39465: Fix _PyUnicode_FromId() for subinterpreters

vstinner opened the pull request:
Make _PyUnicode_FromId() function compatible with subinterpreters.
Each interpreter now has an array of identifier objects (interned
strings decoded from UTF-8).

* Add PyInterpreterState.unicode.identifiers: array of identifiers
  objects.
* Add _PyRuntimeState.unicode_ids used to allocate unique indexes
  to _Py_Identifier.
* Rewrite _Py_Identifier structure.

Benchmark _PyUnicode_FromId(&PyId_a) with _Py_IDENTIFIER(a):

[ref] 2.42 ns +- 0.00 ns -> [atomic] 3.39 ns +- 0.00 ns: 1.40x slower

This change adds 1 ns per _PyUnicode_FromId() call in average.

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-39465](https://bugs.python.org/issue39465) -->
https://bugs.python.org/issue39465
<!-- /issue-number -->

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,74 @@ PyUnicode_FromString(const char *u)
-                                                  NULL, NULL);
-        if (!id->object)
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
```
I would rename the array to `constants`. It can be used not only for strings.

vstinner replied:
What other usage do you forecast? I only plan to use interp->unicode.ids for _PyUnicode_FromId().

serhiy-storchaka replied:
`_PyArg_Parser`

vstinner replied:
Can we rename the member once someone will work on making _PyArg_Parser compatible with subinterpreters? Maybe _PyArg_Parser will use something else.

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,74 @@ PyUnicode_FromString(const char *u)
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
+
+    if (id->index < 0) {
```
I would use 0 as the default value. It is easier to initialize static variables with 0, and it may be slightly faster to compare for equality with 0.

vstinner replied:
"It is easier to initialize static variables with 0"

You cannot initialize a _Py_Identifier to zeros: the string field must be set. IMO it's reasonable to require users of this API to use _Py_static_string_init(), _Py_static_string() or _Py_IDENTIFIER() macro and don't initialize members manually.

serhiy-storchaka replied:
I am not sure, but it looks to me that it may be cheaper for the loader to initialize non-constant static variable with zeros than with some other value. For non-zero value it needs to copy it from some place (read-only) to the read-write data area. For zero value it can just keep initial zeros (if the data page is filled by zeros on allocation). It can save few bytes in executable file and address space and few tacts at program start time.

vstinner replied:
> I am not sure, but it looks to me that it may be cheaper for the loader to initialize non-constant static variable with zeros than with some other value.

_Py_Identifier is made of 2 members: string (non-NULL constant value) and index (set once at runtime, get many times). The structure is never initialized to zeros, even if we change the default "not set" index to 0.

serhiy-storchaka commented on Include/cpython/object.h:
```diff
@@ -36,19 +36,20 @@ PyAPI_FUNC(Py_ssize_t) _Py_GetRefTotal(void);
-    PyObject *object;
+    // Index in PyInterpreterState.unicode.ids.array.
+    // Unique for the whole process.
+    Py_ssize_t index;
```
Perhaps `volatile` is needed now. And maybe a lock or atomic variable. Reading this value will no longer guarded by GIL.

vstinner replied:
Oh right. There is a race condition. I'm not sure what's the most efficient way to assign an unique identifier without hurting performances when the index is already set.

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,74 @@ PyUnicode_FromString(const char *u)
+    if (id->index < 0) {
+        struct _Py_unicode_runtime_ids *rt_ids = &interp->runtime->unicode_ids;
+
+        PyThread_acquire_lock(rt_ids->lock, WAIT_LOCK);
```
Check the value of the index after acquiring the lock. It could be assigned between the previous check and acquiring the lock.

vstinner replied:
Done.

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,74 @@ PyUnicode_FromString(const char *u)
+    }
+
+    PyObject *object;
+    if (id->index < ids->size) {
```
It may be worth to save `id->index` as a local variable. Especially if make it volatile.

vstinner replied:
Done.

vstinner commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,79 @@ PyUnicode_FromString(const char *u)
+        PyThread_acquire_lock(rt_ids->lock, WAIT_LOCK);
+        // Check again to detect concurrent access. Another thread can have
+        // initialized the index while this thread waited for the lock.
+        if (id->index < 0) {
```
I'm not sure if the C compiler must always read again the new structure member. Using volatile avoids the need to ask such question, but I would prefer to avoid volatile if needed.

cc @pablogsal

vstinner replied:
I added volatile. Using atomic variables is not possible right now: pycore_atomic.h remains an internal header, and I don't think that <stdatomic.h> is widely adopted yet.

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
+
+    // Copy the index since _Py_Identifier.index is declared as volatile
+    Py_ssize_t index = id->index;
```
Since reading the index is not guarded by lock, it is possible that we read index simultaneously with writing it in other thread. In such case the half of index bits can be old, and the other half is new. We need not just add `volatile` for index, but use an atomic integer instead of just `Py_ssize_t`.

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
+
+    // Copy the index since _Py_Identifier.index is declared as volatile
+    Py_ssize_t index = id->index;
+    if (index < 0) {
```
Also, I suggest to use 0 as an initial value.

vstinner replied:
See my previous comment: https://github.com/python/cpython/pull/20058#discussion_r424600281

vstinner replied:
_Py_atomic_int_get() is now used.

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
+    assert(index >= 0);
+
+    PyObject *object;
+    if (index < ids->size) {
```
I afraid that `ids->size` should be volatile and atomic too.

vstinner replied:
I chose to avoid atomic variables since pycore_atomic.h is an internal header and I don't think that all C compilers supported by Python implement <stdatomic.h>.

Why does it have to be atomic?

vstinner replied:
If you expect a risk that Py_ssize_t cannot be updated atomatically on all architectures, maybe we can replace it with an int. We have around 500 identifiers currently: it fits into INT_MAX :-)

serhiy-storchaka replied:
It is not guaranteed that `int` or any other type is updated atomically unless you use atomic integers. It is for what they are introduced.

vstinner replied:
`ids->size` is protected by the GIL.

I'm talking about id->index which is volatile but not atomic.

_Py_Identifier.index (id->index) is protected by `_PyRuntime.unicode_ids.lock` which IMO is enough to avoid having to declare _Py_Identifier.index as atomic.

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
+    }
+    PyUnicode_InternInPlace(&object);
+
+    if (index >= ids->size) {
```
The following code should be guarded by a lock too.

vstinner replied:
It's guarded by the GIL (the "interpreter lock").

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
+
+    PyObject *object;
+    if (index < ids->size) {
+        object = ids->array[index];
```
Humm, and the array should be volatile and atomic too if we read it without locking.

vstinner replied:
Again, it's guarded by the GIL.

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
-    return id->object;
+
+    // The array stores a strong reference
+    ids->array[index] = object;
```
There is a possible leak here. In two threads we can simultaneously create an unicode object and save it in the array.

vstinner replied:
It cannot happen because of the GIL, no?

serhiy-storchaka commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
-    return id->object;
+
+    // The array stores a strong reference
+    ids->array[index] = object;
```
There is a possible leak here. In two threads we can simultaneously create an unicode object and save it in the array.

serhiy-storchaka reviewed (commented):
Concurrent programming without GIL is hard.

vstinner reviewed (commented):
Oops, note for myself: I must revert the _testcapi changes, only there for benchmarks.

ericsnowcurrently commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
+        PyThread_release_lock(rt_ids->lock);
+        index = id->index;
+    }
+    assert(index >= 0);
```
I agree with Serhiy about locking around `id->index`.  And if you must lock anyway, why not simplify this section of code (and don't bother with the local `index` var):

```suggestion
    PyThread_acquire_lock(rt_ids->lock, WAIT_LOCK);
    if (id->index) < 0) {
        id->index = rt_ids->next_index;  // The lock guards against concurrent writes (cache ping-pong).
        rt_ids->next_index++;
    }
    PyThread_release_lock(rt_ids->lock);
```

vstinner replied:
I modified my PR to use _Py_atomic_size_get().

ericsnowcurrently commented on Objects/unicodeobject.c:
```diff
@@ -2272,34 +2269,85 @@ PyUnicode_FromString(const char *u)
+    ids->size = 0;
+    PyMem_Free(ids->array);
+    ids->array = NULL;
+    // Don't reset _PyRuntime next_index: _Py_Identifier.id remains valid
```
good comment :slightly_smiling_face:

ericsnowcurrently reviewed (dismissed):
Other than globally-locking around `id->index` before we're sure it's been set, LGTM.

vstinner commented on Objects/unicodeobject.c:
```diff
@@ -2272,40 +2270,83 @@ PyUnicode_FromString(const char *u)
-    return id->object;
+    if (index >= ids->size) {
+        // Overallocate to reduce the number of realloc
+        Py_ssize_t new_size = Py_MAX(index * 2, 16);
```
```suggestion
        Py_ssize_t new_size = Py_MAX((Py_ssize_t)index * 2, 16);
```

vstinner committed:
bpo-39465: Fix _PyUnicode_FromId() for subinterpreters

Make _PyUnicode_FromId() function compatible with subinterpreters.
Each interpreter now has an array of identifier objects (interned
strings decoded from UTF-8).

* Add PyInterpreterState.unicode.identifiers: array of identifiers
  objects.
* Add _PyRuntimeState.unicode_ids used to allocate unique indexes
  to _Py_Identifier.
* Rewrite _Py_Identifier structure.

Benchmark _PyUnicode_FromId(&PyId_a) with _Py_IDENTIFIER(a):

[ref] 2.42 ns +- 0.00 ns -> [atomic] 3.39 ns +- 0.00 ns: 1.40x slower

This change adds 1 ns per _PyUnicode_FromId() call in average.

vstinner committed:
Fix reference leak in subinterpreters
